### PR TITLE
TYLERB/APPEALS-19549: Saving a HLR/SC with incomplete issues

### PR DESCRIPTION
**Pull request by TylerBroyles**
Resolves [APPEALS-19549](https://jira.devops.va.gov/browse/APPEALS-19549)

# Description
The purpose of this PR was to add a way for VHA to save a HLR/SC with incomplete issues.
Some of the steps that were implemented to accomplish this are outlined below:

- Added the necessary backend changes to allow a VHA user to intake a HLR or SC with an issue that does not have a decision date. 
- A method was also added to update the claim's associated DecisionReviewTask's status to on_hold if any of the claim's request issues do not have a decision date. 
- Added a method to change the redirection_url from intake to the decision review queue to the incomplete tab instead of the default tab (in progress) based on the lack of decision date on any of the claim's request issues.
- The wording was also changed for the vha benefit type add issues screen form submission button for HLRs and SCs to reflect these changes. 
- A fix was also added so that the VHA appeals NonratingRequestIssueModal would still require a decision date to submit the modal.

## Acceptance Criteria
- [ ] Code compiles correctly
- [ ] HLR
  - [ ] The intaken review should successfully save and create an on_hold decision review task if an issue has no decision date
  - [ ] The page should navigate to the incomplete tab after intake if an issue has no decision date
  - [ ] The add issues issue page should display Saving on the form submission button instead of Establish if an issue has no decision date
  - [ ] The success message in the decision review queue should match the designs based on the discussion in the vha comms channel
- [ ] SC
  - [ ] The intaken review should successfully save and create an on_hold decision review task if an issue has no decision date
  - [ ] The page should navigate to the incomplete tab after intake if an issue has no decision date
  - [ ] The add issues issue page should display Saving on the form submission button instead of Establish if an issue has no decision date
  - [ ] The success message in the decision review queue should match the designs based on the discussion in the vha comms channel
- [ ] Appeals
  - [ ] The nonrating issue modal should no longer be optional for appeals and should only be optional for HLRs and SCs  

## Testing Plan

1. Go to [Jira Issue/Test Plan Link](https://jira.devops.va.gov/browse/JIRA-12345) or list them below
2. Login as a VHA user that can intake HLRs and SCs
3. Navigate to the intake starting page at the url /intake
4. Start the intake process for an Higher-Level Review
5. Make sure to select the VHA benefit type and advance to the add issues page
6. Add an issue without a decision date
7. Verify that the Save submit button text matches the designs
8. Submit the intaken review
9. Verify that the redirect sends you to the incomplete tab at the url /decision_reviews/vha?tab=incomplete
10. Verify that the success message matches the designs based on the discussion in the vha comms channel
11. Verify that your new claim review is present in the incomplete tab of the VHA decision review queue
12. Verify that when you click on the claimant name of this new task it takes you to the edit issues page
13. Repeat all of the above steps for a Supplemental Claim intake
14. Repeat another intake of either type with multiple issues. Include some which do not have a decision date and some that do. Verify that the behavior is still correct.
15. Repeat another intake for a VHA 10812 appeal
16. Verify that the decision date field on the add issues Nonrating issue type modal is not optional
17. Verify that the modal submit button can not be clicked until a decision date is added for an issue

- [ ] For feature branches merging into master: Was this deployed to UAT?

# Frontend
## User Facing Changes
 - [ ] Screenshots of UI changes added to PR & Original Issue

 BEFORE|AFTER
 ---|---

## Storybook Story
*For Frontend (Presentation) Components*
* [ ] Add a [Storybook](https://github.com/department-of-veterans-affairs/caseflow/wiki/Documenting-React-Components-with-Storybook) file alongside the component file (e.g. create `MyComponent.stories.js` alongside `MyComponent.jsx`)
* [ ] Give it a title that reflects the component's location within the overall Caseflow hierarchy
* [ ] Write a separate story (within the same file) for each discrete variation of the component

# Backend
## Database Changes
*Only for Schema Changes*

* [ ] Add typical timestamps (`created_at`, `updated_at`) for new tables
* [ ] Update column comments; include a "PII" prefix to indicate definite or potential [PII data content](https://github.com/department-of-veterans-affairs/appeals-team/blob/master/caseflow-team/0-how-we-work/pii-handbook.md#what-is-pii)
* [ ] Have your migration classes inherit from `Caseflow::Migration`, especially when adding indexes (use `add_safe_index`) (see [Writing DB migrations](https://github.com/department-of-veterans-affairs/caseflow/wiki/Writing-DB-migrations))
* [ ] Verify that `migrate:rollback` works as desired ([`change` supported functions](https://edgeguides.rubyonrails.org/active_record_migrations.html#using-the-change-method))
* [ ] Perform query profiling (eyeball Rails log, check bullet and fasterer output)
* [ ] For queries using raw sql was an explain plan run by System Team
* [ ] Add appropriate indexes (especially for foreign keys, polymorphic columns, unique constraints, and Rails scopes)
* [ ] Run `make check-fks`; add any missing foreign keys or add to `config/initializers/immigrant.rb` (see [Record associations and Foreign Keys](https://github.com/department-of-veterans-affairs/caseflow/wiki/Record-associations-and-Foreign-Keys))
* [ ] Add `belongs_to` for associations to enable the [schema diagrams](https://department-of-veterans-affairs.github.io/caseflow/task_trees/schema/schema_diagrams) to be automatically updated
* [ ] Document any non-obvious semantics or logic useful for interpreting database data at [Caseflow Data Model and Dictionary](https://github.com/department-of-veterans-affairs/caseflow/wiki/Caseflow-Data-Model-and-Dictionary)

# Best practices
## Code Documentation Updates
- [ ] Add or update code comments at the top of the class, module, and/or component.

### Test Coverage
Did you include any test coverage for your code? Check below:
- [ ] RSpec
- [ ] Jest
- [ ] Other

### Code Climate
Your code does not add any new code climate offenses? If so why?
- [ ] No new code climate issues added

### Monitoring
- [ ] Are performance metrics (e.g., response time, throughput) being tracked?
- [ ] Are key application components monitored (e.g., database, cache, queues)?
- [ ] Is there a system in place for setting up alerts based on performance thresholds?

### Logging
- [ ] Are logs being produced at appropriate log levels (debug, info, warn, error, fatal)?
- [ ] Are logs structured (e.g., using log tags) for easier querying and analysis?
- [ ] Are sensitive data (e.g., passwords, tokens) redacted or omitted from logs?
- [ ] Is log retention and rotation configured correctly?
- [ ] Are logs being forwarded to a centralized logging system if needed?

### Auditing
- [ ] Are user actions being logged for audit purposes?
- [ ] Are changes to critical data being tracked ?
- [ ] Are logs being securely stored and protected from tampering or exposing protected data?

### Error Handling
- [ ] Are errors being caught and handled gracefully?
- [ ] Are appropriate error messages being displayed to users?
- [ ] Are critical errors being reported to an error tracking system (e.g., Sentry, ELK)?
- [ ] Are unhandled exceptions being caught at the application level ?

### Exception Handling
- [ ] Are custom exceptions defined and used where appropriate?
- [ ] Is exception handling consistent throughout the codebase?
- [ ] Are exceptions logged with relevant context and stack trace information?
- [ ] Are exceptions being grouped and categorized for easier analysis and resolution?

### CODE REVIEW

File: spec/feature/intake/vha_hlr_sc_enter_issue_no_decision_date_spec.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+      expect(page).to have_content("Edit Issues")
+      expect(page).to have_content("Decision date: No date entered")
+      expect(page).to have_content(COPY::VHA_NO_DECISION_DATE_BANNER)
+
+      # TODO: Add more stuff after editing is working
+      # Hit the first and add a decision date
+      expect(page).to have_button("Save", disabled: true)
+
+      # within "#issue-0" do
```
Comment (pamatyatake2) [inline spec/feature/intake/vha_hlr_sc_enter_issue_no_decision_date_spec.rb:58]: may be we can remmove this commentted code for now.

File: client/app/intakeEdit/components/EditButtons.jsx [hunk 2]
```diff
@@ -133,7 +134,13 @@ class SaveButtonUnconnected extends React.Component {
       addedIssues, (issue) => issue.withdrawalPending || issue.withdrawalDate
     );
 
-    const saveButtonText = withdrawReview ? COPY.CORRECT_REQUEST_ISSUES_WITHDRAW : COPY.CORRECT_REQUEST_ISSUES_SAVE;
+    let saveButtonText;
+
+    if (benefitType === 'vha' && _.every(addedIssues, (issue) => issue.decisionDate)) {
+      saveButtonText = 'Establish';
```
Comment (pamatyatake2) [inline client/app/intakeEdit/components/EditButtons.jsx:140]: Can we put this 'Establish' to Copy.json for consistency.

File: client/app/intake/components/AddedIssue.jsx [hunk 3]
```diff
@@ -117,7 +117,11 @@ class AddedIssue extends React.PureComponent {
           </div>
         )}
         {issue.benefitType && <span className="issue-date">Benefit type: {BENEFIT_TYPES[issue.benefitType]}</span>}
-        {issue.date && <span className="issue-date">Decision date: {formatDateStr(issue.date)}</span>}
+        {issue.date ? (
+          <span className="issue-date">Decision date: {formatDateStr(issue.date)}</span>
+        ) : (
+          <span className="issue-date">Decision date: No date entered</span>
```
Comment (pamatyatake2) [inline client/app/intake/components/AddedIssue.jsx:123]: May be something like 
<span className="issue-date">Decision date: { issue.date ? formatDateStr(issue.date) : COPY::DECISION_DATE_NO_DATE_ENTERED}</span>

File: app/controllers/claim_review_controller.rb [hunk 4]
```diff
@@ -136,11 +136,34 @@ def review_edited_message
     "You have successfully " + [added_issues, removed_issues, withdrawn_issues].compact.to_sentence + "."
   end
 
+  def vha_edited_decision_date_message
+    "You have successfully updated an issue's decision date"
```
Comment (pamatyatake2) [inline app/controllers/claim_review_controller.rb:140]: may be we can move this string to copy.json, which might help in future if we ever decide to test these messages.

**Comment by pamatyatake2** [comment]
LGTM

**Comment by ajspotts** [comment]
Closed in favor of https://github.com/department-of-veterans-affairs/caseflow/pull/19175

**Pull request closed without merging**